coderd/userauth.go

```diff
@@ -201,6 +202,237 @@ func (api *API) postConvertLoginType(rw http.ResponseWriter, r *http.Request) {
+		ID:                       user.ID,
+		HashedOneTimePasscode:    []byte(hashedPasscode),
+		OneTimePasscodeExpiresAt: sql.NullTime{Time: passcodeExpiresAt, Valid: true},
+	})
+	if err != nil {
+		logger.Error(ctx, "unable to set user hashed one time passcode", slog.Error(err))
+		return
+	}
+
+	newUser := user
+	newUser.HashedOneTimePasscode = []byte(hashedPasscode)
+	newUser.OneTimePasscodeExpiresAt = sql.NullTime{Time: passcodeExpiresAt, Valid: true}
+	aReq.New = newUser
+
+	// Send the one-time-passcode to the user.
+	err = api.notifyUserRequestedOneTimePasscode(ctx, user, passcode.String())
+	if err != nil {
+		logger.Error(ctx, "unable to notify user about one time passcode request", slog.Error(err))
+	}
+}
+
+func (api *API) notifyUserRequestedOneTimePasscode(ctx context.Context, user database.User, passcode string) error {
+	_, err := api.NotificationsEnqueuer.Enqueue(
+		//nolint:gocritic // We need to be able to send the user their one time passcode.
+		dbauthz.AsSystemRestricted(ctx),
+		user.ID,
+		notifications.TemplateUserRequestedOneTimePasscode,
+		map[string]string{"one_time_passcode": passcode},
+		"change-password-with-one-time-passcode",
+		user.ID,
+	)
+	if err != nil {
+		return xerrors.Errorf("enqueue notification: %w", err)
+	}
+
+	return nil
+}
+
+// Change a users password with a one-time-passcode.
+//
+// @Summary Change password with a one-time-passcode.
+// @ID change-password-with-a-one-time-passcode
+// @Accept json
+// @Tags Authorization
+// @Param request body codersdk.ChangePasswordWithOneTimePasscodeRequest true "Change password request"
+// @Success 204
+// @Router /users/change-password-with-one-time-passcode [post]
+func (api *API) postChangePasswordWithOneTimePasscode(rw http.ResponseWriter, r *http.Request) {
+	var (
+		err               error
+		ctx               = r.Context()
+		auditor           = api.Auditor.Load()
+		logger            = api.Logger.Named(userAuthLoggerName)
+		aReq, commitAudit = audit.InitRequest[database.User](rw, &audit.RequestParams{
+			Audit:   *auditor,
+			Log:     api.Logger,
+			Request: r,
+			Action:  database.AuditActionWrite,
+		})
+	)
+	defer commitAudit()
+
+	if api.DeploymentValues.DisablePasswordAuth {
+		httpapi.Write(ctx, rw, http.StatusForbidden, codersdk.Response{
+			Message: "Password authentication is disabled.",
+		})
+		return
+	}
+
+	var req codersdk.ChangePasswordWithOneTimePasscodeRequest
+	if !httpapi.Read(ctx, rw, r, &req) {
+		return
+	}
+
+	//nolint:gocritic // In order to change a user's password, we need to get the user first!
+	user, err := api.Database.GetUserByEmailOrUsername(dbauthz.AsSystemRestricted(ctx), database.GetUserByEmailOrUsernameParams{
+		Email: req.Email,
+	})
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		logger.Error(ctx, "unable to fetch user by email", slog.Error(err))
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error.",
```

Comment:
I want you to imagine you're a user trying to update their password and all you get back in `Internal error.`.

How would an operator go about assisting this user to understand why their request failed? Do they have sufficient context in the logs to isolate this user (amongst possibly thousands) who are experiencing some issue?

Reply:
This flow should (from my understanding) only be hit if there is an underlying issue with the database. In that case I'm assuming this would be less of an individual user issue but instead a more general issue.

If we do want to add more context the only information at that point we could use  is the given user email address?

Reply:
Yup, that should be sufficient.